### PR TITLE
Add Asttypes.constant abstractions to ppx_tools.

### PR DESCRIPTION
**Pull request by neolem**
@Drup This pull request adds a new function `constant_type` to abstract ocaml AST const types. This is done in relation to and as an offshoot of discussion at [sedlex pull request](https://github.com/alainfrisch/sedlex/pull/37). The latter pull request depends on this pull request. I am sending another pull request for ppx_tools 4.03 branch(master).

### CODE REVIEW

**Comment by Drup** [comment]
Hum, don't we want to do the opposite: Translate to the newest version ? 
Also, I think it should be in both direction, like in the `Label` module.

**Comment by alainfrisch** [comment]
We should discourage using `Asttypes.constant` in syntactic tools since it looses information compared to the new `Parsetree.constant` (one could argue that `Asttypes.constant` should be moved to a module in the typing/ subdirectory).

What about exposing (for the 4.02 compat layer) a copy of the new more expressive type and functions to map to/from the one used in Parsetree?  For 4.03,  one would simply re-export the actual new type and provide identity functions.

**Comment by Drup** [comment]
> What about exposing (for the 4.02 compat layer) a copy of the new more expressive type and functions to map to/from the one used in Parsetree? For 4.03, one would simply re-export the actual new type and provide identity functions.

I agree this direction is preferable.

**Comment by neolem** [comment]
Indeed, I was thinking along the same lines but thought best not to make too many changes. I will update the pull requests.

**Comment by neolem** [comment]
@Drup Updated the pull requests with required changes. Ready for your review.

**Comment by Drup** [comment]
@neolem Nice, Could you add the reversed function, `t -> constant` ?

File: ast_convenience.mli [hunk 1]
```diff
@@ -86,6 +98,8 @@ val get_str: expression -> string option
 val get_str_with_quotation_delimiter: expression -> (string * string option) option
 val get_lid: expression -> string option
 
+(** {AST deconstruction functions for Asttypes.Constant types} *)
```
Comment (alainfrisch) [inline ast_convenience.mli]: This comment seems to be irrelevant.

File: ast_convenience.mli [hunk 2]
```diff
@@ -26,11 +26,23 @@ module Label : sig
...
+  (** Translate ocaml version specific constant type to Constant.t *)
+  val constant_type : constant -> t
+end
+
 (** {2 Misc} *)
 
 val lid: string -> lid
 
-(** {2 Expressions} *)
+(** {2 Expressions. } *)
```
Comment (alainfrisch) [inline ast_convenience.mli]: Other section titles don't have a trailing dot.

**Comment by neolem** [comment]
@Drup @alainfrisch I have checked in a couple of commits to address the review points. Ready for review.

File: ast_convenience.ml [hunk 3]
```diff
@@ -28,6 +28,32 @@ module Label = struct
...
+    | Asttypes.Const_int64(i) -> Pconst_integer(Int64.to_string i, Some 'L')
+    | Asttypes.Const_nativeint(i) -> Pconst_integer(Nativeint.to_string i, Some 'n')
+    | Asttypes.Const_int(i) -> Pconst_integer(string_of_int i, None)
+    | Asttypes.Const_char c -> Pconst_char c 
+    | Asttypes.Const_string(s, s_opt) -> Pconst_string(s, s_opt) 
+    | Asttypes.Const_float f -> Pconst_float(f, None)
+
+  let to_constant = function 
+    | Pconst_integer(s, Some 'l') -> Asttypes.Const_int32 (Int32.of_string s)
+    | Pconst_integer(s, Some 'L') -> Asttypes.Const_int64 (Int64.of_string s)
```
Comment (hhugo) [inline ast_convenience.ml]: This is probably not 100% correct, see what's done in the compiler
https://github.com/ocaml/ocaml/blob/trunk/utils/misc.ml#L298

**Comment by alainfrisch** [comment]
Thinking more generally about how to write ppx (or other syntactic tools) compatible across multiple versions of OCaml without introducing a lot of noise in their code, it might be worth mapping the Parsetree globally.  Meaning ppx_tools (or another package) could expose a copy of the Parsetree types from the latest OCaml version and functions to map to/from that type and the one for the supported version.  The direction from the "new" Parsetree to the old one would be partial: it can fail if the ppx inserts new language features not compatible with the old OCaml; but this should rarely matter in practice.  One would also provide copies ot the "current" Ast_helper and Ast_mapper modules from the current OCaml.  This approach (which generalizes what has been done in this ticker to the entire Parsetree) means that the code of the ppx would not be cluttered with "local" mapping.

It's a lot of tedious work, but people who care about supporting old versions of OCaml might find this useful.

The only constraint to make this possible is that the current Parsetree definition remains valid OCaml code for supported old versions of OCaml.  (For instance, if we want to go back before the introduction of GADTs, this feature should not be used in the Parsetree.  Luckily, we are usually rather conservative in the use of recent features in the compiler code base, and the Parsetree has no reason to use exotic new features -- except perhaps inline records and/or type-based selection of fields/constructors.)

File: ast_convenience.ml (hunk 3 above)
Comment (neolem) [inline ast_convenience.ml]: @hhugo :+1:. Thank you for the pointer. Indeed 4.03 has more robust `int{32,64,nativeint}` conversion functionality. I have back-ported this to `ppx_tools` 4.02 branch.

**Comment by neolem** [comment]
@alainfrisch whao ... this would be quite a handful. Indeed, this would be a big chunk of initial and ongoing work to maintain such functionality.

**Comment by alainfrisch** [comment]
Thanks!
